include/ginkgo/core/base/types.hpp

```diff
@@ -89,11 +89,26 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 
 // Handle deprecated notices correctly on different systems
-#if defined(_WIN32) || defined(__CYGWIN__)
-#define GKO_DEPRECATED(msg) __declspec(deprecated(msg))
+// clang-format off
+#define GKO_DEPRECATED(_msg) [[deprecated(_msg)]]
```

Comment:
Since deprecations are a language feature, I would prefer if we use them directly, instead of this macro. (I didn't even know that we have this macro before).

Reply:
same commet if [[deprecated()]] is also allowed in WIN32 and CYGWIN

Reply:
By using a macro, we make it easier to make the warnings configurable (e.g. a config.hpp switch that disables deprecation warnings?), but I'm not particularly attached to that change. Only we probably want to keep the macro around for interface stability.

Reply:
For now, I decided to leave the macro in place without an option to disable it. That means we just use the C++14 method and still have the option to disable it in the future if we want to (but, as far as I can see, we don't emit any false-positives anymore),
We do, however, still emit CUDA and HIP deprecation warnings for some compilers, but no functions that are marked as deprecated inside Ginkgo.